tools/c7n_azure/c7n_azure/resources/access_control.py

```diff
@@ -96,3 +141,88 @@ class UserRole(RelatedResourceFilter):
 
     RelatedResource = "c7n_azure.resources.access_control.RoleDefinition"
     RelatedIdsExpression = "properties.roleDefinitionId"
+
+
+@RoleAssignment.filter_registry.register('resource-access')
+class ResourceAccessFilter(Filter):
+    """Filters role assignments that have access to a certain
+    type of azure resource.
+
+    :Example:
+
+        .. code-block:: yaml
+
+            policies:
+               - name: assignments-by-azure-resource
+                 resource: azure.roleassignment
+                 filters:
+                    - type: resource-access
+                      relatedResource: azure.vm
+    """
+
+    schema = type_schema(
+        'resource-access',
+        relatedResource={'type': 'string'},
+        required=['relatedResource']
+    )
+
+    def __init__(self, data, manager=None):
+        super(ResourceAccessFilter, self).__init__(data, manager)
+        resource_type = self.data['relatedResource']
+        self.factory = Azure.resources.get(
+            resource_type.rsplit('.', 1)[-1])
+
+    def get_related(self):
+        ctx = ExecutionContext(local_session(Session), self.data, Config.empty())
+
+        manager = self.factory(ctx, self.data)
+        return manager.source.get_resources(None)
+
+    def process(self, resources, event=None):
+        related_resources = self.get_related()
+        s = local_session(self.manager.session_factory)
+        assignment_ids = []
+        headers = {
+            'Content-type': 'application/json',
+            'Authorization': 'Bearer %s' % (s.get_bearer_token())
+        }
+        for resource in related_resources:
+            url = (
+                'https://management.azure.com%s/'
+                'providers/Microsoft.Authorization/roleAssignments?api-version=2015-07-01'
+                % (resource['id'])
+            )
+            response = requests.get(url, headers=headers)
+            results = response.json()
+            assignment_ids.extend([r['id'] for r in results['value']])
+
+        assignment_ids = list(set(assignment_ids))
+        return [r for r in resources if r['id'] in assignment_ids]
```

Comment:
Not sure if I'm following, the example shows you use this filter on a policy of type `azure.roleassignment` which implies `resources` is already the global list of assignments, but then in `Process` you are getting new role assignments and filtering them by themselves? Shouldn't we already have all of them and we should just be enumerating VM's here?  A normal related resource filter might work OK filtering on whether the resource ID key contains the scope field?

Reply:
So you are suggesting we accept azure.vm as the resource array and then return role assignments? It's possible but it goes pretty much against all other filters in that they filter the provided resource array and return the filtered array (filtering on role assignments returns assignments, filtering on vms returns vms).

Reply:
The first 2 paragraphs I understand. Could just check if the resource ID contains the scope, which would work! I don't think scope can be arrays.

Reply:
Oh yeah I was referring to inverting into a more traditional value filter, I think you probably planned for this to happen additionally later anyway.  e.g. Policy on a VM, related item filter on Assignment, normal value filter stuff to filter VM's by the assignment details so you just end up with VM's that have some certain type of assignment (or lack thereof).

Reply:
Best I can tell related resource filter uses valuefilter matching to decide if it is related - meaning you can do `in` and `contains` ops, etc between the two keys- although I also though you could additionally do filtering on other properties of the related resource - I'll have to dig into that one a little more.

Reply:
Yeah that was my next PR. But doing the related resource on the scope and resource key might be way easier both ways now. Except the actual related resource filter requires an ID to link them and there is still no direct ID. But definitely worth investigating getting a "contains" related ID to work out.